lib/src/streams/data_item.dart

```diff
@@ -0,0 +1,176 @@
+DataItemTaskEither createDataItemTaskEither({
+  required final Wallet wallet,
+  required final DataStreamGenerator dataStream,
+  required final int dataStreamSize,
+  final String target = '',
+  final String anchor = '',
+  final List<Tag> tags = const [],
+}) {
+  return getOwnerTaskEither(wallet).flatMap((owner) =>
+      decodeBase64ToBytesTaskEither(owner).flatMap((ownerBytes) =>
+          decodeBase64ToBytesTaskEither(target).flatMap((targetBytes) {
+            if (target.isNotEmpty && targetBytes.length != 32) {
+              return TaskEither.left(InvalidTargetSizeError());
+            }
+
+            return decodeBase64ToBytesTaskEither(anchor).flatMap((anchorBytes) {
+              if (anchor.isNotEmpty && anchorBytes.length != 32) {
+                return TaskEither.left(InvalidAnchorSizeError());
+              }
+
+              final tagsBytes = serializeTags(tags: tags);
+
+              return deepHashTaskEither([
+                toStream(utf8.encode('dataitem')),
+                toStream(utf8.encode('1')), // Transaction format
+                toStream(utf8.encode('1')),
+                toStream(ownerBytes),
+                toStream(targetBytes),
+                toStream(anchorBytes),
+                toStream(tagsBytes),
+                dataStream(),
+              ]).flatMap((signatureData) => signDataItemTaskEither(
+                          wallet: wallet, signatureData: signatureData)
+                      .flatMap((signResult) {
+                    final dataItemHeaders = [
+                      ...shortTo2ByteArray(1),
+                      ...signResult.signature,
+                      ...ownerBytes,
+                      ...(targetBytes.isEmpty ? [0] : [1]),
+                      ...targetBytes,
+                      ...(anchorBytes.isEmpty ? [0] : [1]),
+                      ...anchorBytes,
+                      ...longTo8ByteArray(tags.length),
+                      ...longTo8ByteArray(tagsBytes.lengthInBytes),
+                      ...tagsBytes,
+                    ];
+
+                    final dataItemStreamGenerator =
+                        combineStreamAndFunctionList(
+                            toStream(dataItemHeaders), [dataStream]);
+
+                    return TaskEither.of(DataItemResult(
+                      id: signResult.id,
+                      dataItemSize: dataItemHeaders.length + dataStreamSize,
+                      dataSize: dataStreamSize,
+                      streamGenerator: dataItemStreamGenerator,
+                    ));
+                  }));
+            });
+          })));
+}
+
+Future<ProcessedDataItem> processDataItem({
+  required Stream<List<int>> Function() dataItemStreamGenerator,
+  required String id,
+  required int length,
+}) async {
+  int byteIndex = 0;
+  final reader = ChunkedStreamReader(dataItemStreamGenerator());
+
+  // get signature type
+  final signatureType = decodeBytesToLong(await reader.readBytes(2));
+  byteIndex += 2;
```

Comment:
Trying to understand whats going on here - so its reading the bytes of a stream in a particular order, where the first 2 bytes are the signature type, followed by the sig, owner, etc?